core/src/main/java/com/datastrato/gravitino/storage/kv/KvGarbageCollector.java

```diff
@@ -232,6 +253,71 @@ private void collectAndRemoveOldVersionData() throws IOException {
         kvBackend.delete(kv.getKey());
       }
     }
+
+    commitIdHasBeenCollected = kvs.isEmpty() ? startKey : kvs.get(0).getKey();
+    kvBackend.put(LAST_COLLECT_COMMIT_ID_KEY, commitIdHasBeenCollected, true);
+  }
+
+  /**
+   * Remove all versions of the key.
+   *
+   * @param rawKey raw key, it contains the transaction id.
+   * @param key key, it's the real key and does not contain the transaction id
+   * @param includeStart whether include the start key.
+   * @throws IOException if an I/O exception occurs during deletion.
+   */
+  private void removeAllVersionsOfKey(byte[] rawKey, byte[] key, boolean includeStart)
+      throws IOException {
+    List<Pair<byte[], byte[]>> kvs =
+        kvBackend.scan(
+            new KvRange.KvRangeBuilder()
+                .start(rawKey)
+                .end(generateKey(key, 1))
+                .startInclusive(includeStart)
+                .endInclusive(false)
+                .build());
+
+    for (Pair<byte[], byte[]> kv : kvs) {
+      // Delete real data.
+      kvBackend.delete(kv.getKey());
+
+      LogHelper logHelper = decodeKey(kv.getKey());
+      LOG.info(
+          "Physically delete key that has marked deleted: name identifier: '{}', entity type: '{}', createTime: '{}({})', key: '{}'",
+          logHelper.identifier,
+          logHelper.type,
+          logHelper.createTimeAsString,
+          logHelper.createTimeInMs,
+          Bytes.wrap(key));
+
+      // Try to delete commit id if the all keys in the transaction id have been dropped.
+      byte[] transactionId = getBinaryTransactionId(kv.getKey());
+      byte[] transactionKey = generateCommitKey(transactionId);
+      byte[] transactionValue = kvBackend.get(transactionKey);
+
+      List<byte[]> keysInTheTransaction = SerializationUtils.deserialize(transactionValue);
+
+      boolean allDropped = true;
+      for (byte[] keyInTheTransaction : keysInTheTransaction) {
+        if (kvBackend.get(generateKey(keyInTheTransaction, transactionId)) != null) {
+          // There is still a key in the transaction, we cannot delete the commit mark.
+          allDropped = false;
+          break;
+        }
+      }
+
+      // Try to delete the commit mark.
+      if (allDropped) {
+        long timestamp = TransactionalKvBackendImpl.getTransactionId(transactionId) >> 18;
+        LOG.info(
+            "Physically delete commit mark: {}, createTime: '{}({})', key: '{}'",
+            Bytes.wrap(kv.getKey()),
+            DateFormatUtils.format(timestamp, TIME_STAMP_FORMAT),
+            timestamp,
+            Bytes.wrap(kv.getKey()));
+        kvBackend.delete(transactionKey);
+      }
+    }
```

Comment:
What will be happened here if any of the storage operation is failed (scan, get, delete)?

Reply:
If any steps fails, it throws an exceptions and we would repeated do it in the next time, considering the following scenario:
-1 Scan transactions id
-2 For each transaction ID, check the data in the transaction.
-3 Drop the data if it needs to be deleted( deleted or with a newer version)
-4 Remove the transaction marks. 
-5 done.

For any failures from steps 1 to 5,  do `collectAndRemoveOldVersionData` again will solve it as the value of `commitIdHasBeenCollected` will not move forward if anything unexpected happens.